### ISSUE TITLE
Silent memory corruption when custom_mask shape doesn't match (qo_len, kv_len) in single_prefill_with_kv_cache

### ISSUE THREAD

**Issue by frauttauteffasu** [flashinfer-ai/flashinfer#3176]
While vllm-omni was adding support for flashinfer a [data corruption/illegal access] issue was detected due to lack of validation of the `custom_mask`'s shape. The [functions documentation](https://docs.flashinfer.ai/generated/flashinfer.prefill.single_prefill_with_kv_cache.html) states
> custom_mask (Optional[torch.Tensor]) –
>
> The custom boolean mask tensor, shape: [qo_len, kv_len]. The elements in the mask tensor should be either True or False, where False means the corresponding element in the attention matrix will be masked out.

I am unable to provide a simple reproducer for illegal access. @lishunyang12 was able to produce it on end to end test of Hunyuan Video 1.5 with https://github.com/vllm-project/vllm-omni/pull/3079/changes/b25c755322cba84e14802e1f62acf5f982480618. A simpler reproducer that detects corruption is:
```
import torch
import flashinfer


def test_bad_mask_shape():
    qo_len, kv_len = 10, 20
    num_qo_heads, num_kv_heads = 8, 4
    head_dim = 128
    kwargs= {}

    q = torch.randn(qo_len, num_qo_heads, head_dim, dtype=torch.float16)
    k = torch.randn(kv_len, num_kv_heads, head_dim, dtype=torch.float16)
    v = torch.randn(kv_len, num_kv_heads, head_dim, dtype=torch.float16)

    bad_mask = torch.ones(qo_len, kv_len - 5, dtype=torch.bool)

    # Correct mask for comparison
    correct_mask = torch.ones(qo_len, kv_len, dtype=torch.bool)
    # Mask out the last 5 columns to make it meaningful
    correct_mask[:, 15:] = False

    print(f"Q shape: {q.shape} (qo_len={qo_len})")
    print(f"K shape: {k.shape} (kv_len={kv_len})")
    print(f"V shape: {v.shape}")
    print(f"Bad mask shape: {bad_mask.shape} (expected [{qo_len}, {kv_len}])")
    print(f"Correct mask shape: {correct_mask.shape}")
    print(f"Bad packed mask size: {(bad_mask.numel() + 7) // 8} bytes")
    print(f"Correct packed mask size: {(correct_mask.numel() + 7) // 8} bytes")
    print()

    print("Running with CORRECT mask...")
    kwargs["custom_mask"] = correct_mask
    o_correct = flashinfer.single_prefill_with_kv_cache(q, k, v, ** kwargs)
    print(f"Correct output (first head, first 5 tokens): {o_correct[0, 0, :5]}")

    # Run with bad mask
    print("\nRunning with BAD mask (shape mismatch)...")
    try:
        kwargs["custom_mask" = bad_mask
        o_bad = flashinfer.single_prefill_with_kv_cache(q, k, v, ** kwargs)
        print(f"Bad output (first head, first 5 tokens): {o_bad[0, 0, :5]}")
        
        # Check for corruption
        diff = (o_correct - o_bad).abs().max().item()
        print(f"\nMax absolute difference: {diff:.6f}")
        
        if diff > 1e-5:
            print("Results differ significantly!")
            
    except RuntimeError as e:
        print("This is the kernel detecting the illegal access.")


if __name__ == "__main__":
    test_bad_mask_shape()
```
